File: haskell/anagram/src/Anagram.hs
module Anagram (anagramsFor) where

import Data.List (sort)
import Data.Char (toLower)

isAnagram :: String -> String -> Bool
isAnagram x y =
     x_        /= y_
  && length x_ == length y_
  && sort x_   == sort y_
  where
    x_ = map toLower x
    y_ = map toLower y

anagramsFor :: String -> [String] -> [String]
anagramsFor xs = filter (isAnagram xs)

File: javascript/anagram/anagram.js
const normalizeString = x =>
  JSON.stringify(
    [...x.toLowerCase()]
      .sort()
  )

const isAnagram = x => y =>
	x !== y &&
  x.toLowerCase() !== y.toLowerCase() &&
	x.length === y.length &&
  normalizeString(x) === normalizeString(y)

export default class Anagram {
	constructor(string) {
		this.string = string
	}

	matches(list) {
		return list.filter(isAnagram(this.string))
	}
}
